cwltool/pathmapper.py

```diff
@@ -171,7 +171,9 @@ def setup(self, referenced_files: List[CWLObjectType], basedir: str) -> None:
         stagedir = self.stagedir
         for fob in referenced_files:
             if self.separateDirs:
-                stagedir = os.path.join(self.stagedir, "stg%s" % uuid.uuid4())
+                stagedir = os.path.join(self.stagedir,
+                                        "s5g%s" % uuid.uuid5(uuid.UUID('6a56ca02-b6f0-4c1a-a4b0-fb0068ce80ad'),
```

Comment:
???

Reply:
uuid5 is deterministic.  This groups together files that were previously grouped together by providing the same uuid for the each file with the same basedir, so it violates completely separating files.  This _should_ avoid conflicts.

Reply:
ok! Why `s5g` instead of `stg`?

Reply:
Shouldn't you be passing something to `uuid.uuid5` related to the `fob.location`?

Reply:
I it changed `s5g` just in case this needs to distinguish behavior in the future based on versions.  Though it's not necessary.

And `fob['location']` is line 176, the comment blurb doesn't make it visible here so you'll have to look at the diff.

Reply:
So `"s5g%s" % uuid.uuid5(uuid.UUID('6a56ca02-b6f0-4c1a-a4b0-fb0068ce80ad'` generates a static string that will never change, to which you make a subdir based upon the directory name of the `location` field. So why not skip to the directory name itself? Or use it to see the `uuid5` generator?

What if there are location URIs that end in the same directory name but are otherwise unrelated? This could lead to files overwriting each other. Or files being adjacent to one another when they should not be.

Reply:
I'll go through my thinking on this, and please point out any flaws.

Uuid5 generates different uuids for different seeds, the seed in this PR being both the file's directory and a static uuid.  So `uuid.uuid5(arg1, arg2) = uuid_a` and `uuid.uuid5(arg1, arg3) = uuid_b`.  The different is that if `uuid.uuid5(arg1, arg2) = uuid_a` occurs again, it will generate the same `uuid_a` output.

```
>>> from uuid import uuid5, UUID
>>> uuid5(UUID('6a56ca02-b6f0-4c1a-a4b0-fb0068ce80ad'), '/tmp/1')
UUID('e3e49c59-338c-5369-a215-3d68bda76d69')
>>> uuid5(UUID('6a56ca02-b6f0-4c1a-a4b0-fb0068ce80ad'), '/tmp/2')
UUID('48220b67-6c36-500b-ae08-ecfa48951864')
>>> uuid5(UUID('6a56ca02-b6f0-4c1a-a4b0-fb0068ce80ad'), '/tmp')
UUID('98da1ac5-114d-539b-b727-948e7754ec95')
>>> uuid5(UUID('6a56ca02-b6f0-4c1a-a4b0-fb0068ce80ad'), '/tmp/1')
UUID('e3e49c59-338c-5369-a215-3d68bda76d69')
```

The idea being that all pre-existing paths cannot conflict since they co-existed together peacefully in the first place.

So files:

```
/tmp/a/b.txt
/tmp/b/a.txt
/tmp/b/b.txt
/tmp/c/b.txt
```

would map to these inside of the container:

```
uuid_a/b.txt
uuid_b/a.txt
uuid_b/b.txt
uuid_c/b.txt
```

> So why not skip to the directory name itself?

If I understand your question correctly, we could use the same directory names as the files originally had, but they might conflict with folders that already exist within the container.

> Or use it to see the uuid5 generator?

If I understand this to be to "seed" the uuid5 generator with the directory, I'm using `os.path.dirname(fob['location']))` to seed the uuid5.  Though I'm not sure if that's the question here.

> What if there are location URIs that end in the same directory name but are otherwise unrelated? This could lead to files overwriting each other. Or files being adjacent to one another when they should not be.

I was wondering about this, and was trying to reason if this could happen if the input files were writable in a an input dir outside of the workdir.  I haven't tested this.

Some of this too is that I tried to switch to a simpler strategy and it may just not work.